internal/pkg/gateway/api.go

```diff
@@ -173,183 +169,190 @@ func (gs *Server) Endorse(ctx context.Context, request *gp.EndorseRequest) (*gp.
+	if err != nil {
+		return nil, status.Errorf(codes.Aborted, "failed to assemble transaction: %s", err)
+	}
+
+	return &gp.EndorseResponse{PreparedTransaction: preparedTransaction}, nil
+}
+
+type ppResponse struct {
+	response *peer.ProposalResponse
+	err      error
+}
+
+func (gs *Server) processProposal(ctx context.Context, plan *plan, endorser *endorser, signedProposal *peer.SignedProposal, logger *flogging.FabricLogger) (*endorser, bool) {
+	var response *peer.ProposalResponse
+	done := make(chan *ppResponse)
+	go func() {
+		defer close(done)
+		logger.Debugw("Sending to endorser:", "MSPID", endorser.mspid, "endpoint", endorser.address)
+		ctx, cancel := context.WithTimeout(ctx, gs.options.EndorsementTimeout) // timeout of individual endorsement
+		defer cancel()
+		response, err := endorser.client.ProcessProposal(ctx, signedProposal)
+		done <- &ppResponse{response: response, err: err}
+	}()
+	select {
+	case resp := <-done:
+		// Endorser completed normally
+		code, message, _, remove := responseStatus(resp.response, resp.err)
+		if code != codes.OK {
+			logger.Warnw("Endorse call to endorser failed", "MSPID", endorser.mspid, "endpoint", endorser.address, "error", message)
+			if remove {
+				gs.registry.removeEndorser(endorser)
 			}
+			plan.addError(errorDetail(endorser.endpointConfig, message))
+			return plan.nextPeerInGroup(endorser), false
 		}
+		response = resp.response
+		logger.Debugw("Endorse call to endorser returned success", "MSPID", endorser.mspid, "endpoint", endorser.address, "status", response.Response.Status, "message", response.Response.Message)
 
-		// 5. Get a set of endorsers from discovery via the registry
-		// The preferred discovery layout will contain the firstEndorser's Org.
-		plan, err = gs.registry.endorsementPlan(channel, interest, firstEndorser)
-		if err != nil {
-			return nil, status.Error(codes.FailedPrecondition, err.Error())
+		responseMessage := response.GetResponse()
+		if responseMessage != nil {
+			responseMessage.Payload = nil // Remove any duplicate response payload
 		}
 
-		// 6. Remove the gateway org's endorser, since we've already done that
-		action, err = plan.processEndorsement(firstEndorser, firstResponse)
-		if err != nil {
-			return nil, status.Error(codes.Aborted, err.Error())
+		success := plan.processEndorsement(endorser, response)
+		if success {
+			return nil, true
 		}
+		return plan.nextPeerInGroup(endorser), false
+	case <-ctx.Done():
+		// Overall endorsement timeout expired
+		return nil, false
 	}
+}
 
-	var errorDetails []proto.Message
-	for action == nil {
-		// loop through the layouts until one gets satisfied
-		endorsers := plan.endorsers()
-		if endorsers == nil {
-			// no more layouts
-			break
+func (gs *Server) planFromFirstEndorser(ctx context.Context, channel string, chaincodeID string, hasTransientData bool, signedProposal *peer.SignedProposal, logger *flogging.FabricLogger) (*plan, error) {
```

Comment:
For these non-trivial functions, it helps to have a function comment explaining what the function does.